Model/Serverside/Request/Frontend.php

```diff
@@ -0,0 +1,274 @@
+        ServersideHelper $serverSide,
+        ProductData $productData,
+        Cookie $cookie,
+        General $general,
+        Store $store,
+        Cart $cart,
+        Customer $customer
+    ) {
+        $this->requestFactory = $requestFactory;
+        $this->serverSide     = $serverSide;
+        $this->productData    = $productData;
+        $this->cookie         = $cookie;
+        $this->general        = $general;
+        $this->store          = $store;
+        $this->cart           = $cart;
+        $this->customer       = $customer;
+    }
+
+    /**
+     * @param string $storeId
+     */
+    public function setStoreId($storeId)
+    {
+        $this->storeId = $storeId;
+    }
+
+    /**
+     * Builds and sends a serverside request
+     *
+     * @param mixed[] $params
+     * @return mixed[]
+     */
+    public function execute($params)
+    {
+        $request = $this->buildRequest($params);
+        $this->serverSide->dispatch($request);
+        return $this->processResult($request, $this->storeId);
+    }
+
+    /**
+     * Builds the request information into the request object
+     *
+     * @param mixed[] $params
+     * @return Data
+     */
+    public function buildRequest($params)
+    {
+        /** @var Data $request */
+        $this->request = $this->requestFactory->create();
+
+        $this->addGeneralInfo($this->storeId, $params);
+        $this->addZones($params);
+        $this->addPageView($params);
+        $this->addProductView($params);
+        $this->addCustomerDetails($this->storeId);
+        $this->addSetBasket();
+        $this->addOrder($params);
+
+        return $this->request;
+    }
+
+    /**
+     * Adds the General information to the Request (App info, urls, currency etc)
+     *
+     * @param string $storeId
+     * @param mixed[] $params
+     */
+    public function addGeneralInfo($storeId, $params)
+    {
+        $this->serverSide->setStoreId($storeId);
+
+        $this->request->setAppId($this->store->getAccessKey($storeId));
+        $this->request->setSecretKey($this->store->getSecretKey($storeId));
+        $this->request->setUserAgent($this->general->getUserAgent());
+        $this->request->setIp($this->general->getIp());
+        $this->request->setCurrency($this->store->getCurrency($storeId));
+        $this->request->setCurrentUrl(isset($params['current_url']) ? $params['current_url'] : '');
+        $this->request->setReferer(isset($params['referer']) ? $params['referer'] : '');
+    }
+
+    /**
+     * Adds specified Zones to the request
+     *
+     * @param mixed[] $params
+     */
+    public function addZones($params)
+    {
+        if (isset($params['zones'])) {
+            foreach ($params['zones'] as $zone) {
+                $this->request->addZone($zone, []);
+            }
+        }
+    }
+
+    /**
+     * Adds Page View event to the request
+     *
+     * @param mixed[] $params
+     */
+    public function addPageView($params)
+    {
```

Comment:
did we discuss setting the page_type?

Reply:
Just thinking about the [page contexts](https://www.pureclarity.com/docs/bespoke-integration/#overview_5). Do you currently set the category/brand context per zone?

Reply:
Oh, and whilst I think about contexts - do you record the search term?

Reply:
page_type is built in getPageContext in  ViewModel/Configuration.php. 

We don't do it per zone, but it is sent in the page view / product view context

will add search term context